src/psyclone/psyir/transformations/arrayrange2loop_trans.py

```diff
@@ -214,6 +238,31 @@ def validate(self, node, options=None):
                             f"assignment '{node}'.")
                     break
 
+        if not options:
+            options = {}
+        allow_string_array = options.get("allow_string", False)
+        # If we allow string arrays then we can skip the check.
+        if not allow_string_array:
+            # ArrayMixin datatype lookup can fail if the indices contain a
+            # Call or Intrinsic Call. We catch this exception and continue
+            # for now - TODO #1799
+            for child in node.walk((Literal, Reference)):
+                try:
+                    # Skip unresolved types
+                    if isinstance(child.datatype,
+                                  (UnresolvedType, UnsupportedType)):
+                        continue
```

Comment:
This line is also missing from the coverage report, the same line for NEMO is not missing so hopefuly is only a matter to copy over one of the test cases to the generic version. (keep it the same if possible as this have to be merged soon into a single implementation and will be easier to do if we know exactly what is duplicated)

Reply:
I'm very unclear which test is covering this for the Nemo version.

I expected it to be the `character(LEN=x` test but this still gives a `child.datatype` of ArrayType I think which is not causing the `continue` to be taken, This actually comes from a test structure of ararys apparently...

Reply:
Which makes me feel this code isn't doing what we want it to do. I think we need a longer if statement:
```

Reply:
```                   
 if (isinstance(child.datatype,
                                   (UnresolvedType, UnsupportedType)) or
                        (isinstance(child.datatype, ArrayType) and
                         isinstance(child.datatype.datatype,
                             (UnresolvedType, UnsupportedType))))
```